doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,334 @@
+3. Spawn a new `tokio::task`, do all further work in this task and off the `Coordinator` thread.
+4. Decode the body of our request according to `BODY FORMAT`.
+5. Pack our data into a `Row`.
+6. Send a tuple of `(Row, 1)` to the `StorageController`, wait for it the data to committed.
+7. Use the provided `oneshot::Sender` to respond with success or failure.
+
+Today we expose a method [`append`](https://github.com/MaterializeInc/materialize/blob/149d2e5bca1f54e94acfe02ab5f5751b9cb2adfc/src/storage-client/src/controller.rs#L403-L406)
+on the `StorageController` trait, I would like to add a new method:
+
+```
+pub async fn monotonic_appender(&self, id: GlobalId) -> CollectionAppender;
+```
+
+This new method would return a struct called `CollectionAppender`, which essentially acts as a
+`oneshot::Sender<Vec<(Row, Diff)>>`. It's a `Send`-able and non-`Clone`-able struct that has one
+method `async fn append(self, ...)` where it takes ownership of `self` so it can only be used once.
+This would be implemented as part of [`collection_mgmt`](https://github.com/MaterializeInc/materialize/blob/main/src/storage-client/src/controller/collection_mgmt.rs)
+and would approximately be structed like:
+
+```
+struct CollectionAppender {
+  id: GlobalId,
+  tx: mpsc::Sender<(GlobalId, Vec<(Row, Diff)>)>,
+}
+```
+
+In addition to the `CollectionAppender`, we'll update the `CollectionManager` to receive a
+`oneshot::Sender<Result<(), StorageError>>` that it can use to optionally notify any listeners when
+and if the updates were successfully committed.
+
+Adding the `CollectionAppender` introduces a possible correctness issue though:
+
+1. Create a `CollectionAppender` for `GlobalId(A)`.
+2. Drop source associated with `GlobalId(A)`.
+3. Try to append data.
+
+To protect against this we can change the `CollectionManager` to return an error via the
+`oneshot::Sender` if the [provided GlobalId no longer exists](https://github.com/MaterializeInc/materialize/blob/main/src/storage-client/src/controller/collection_mgmt.rs#L83).
+
+Adding this new API gets us two important properties that the existing `append` API does not have:
+
+1. The ability to send updates to the `StorageController` from a thread other than the main
+   `Coordinator` thread.
+2. Allows the `StorageController` to pick the timestamp for the update, instead of the
+   `Coordinator`, which aligns with our long term goals if having a storage owned "push source" in
+   the future.
+
+### Request Validation
+[request_validation]: #request_validation
+
+Every `WEBHOOK` source will be open to the entire internet, as such we need to include some sort of
+validation so we can verify that requests are legitimate. This is can be fairly difficult to do
+though because it could require us to run arbitrary logic when receiving each event, e.g. something
+very similar to `TRANSFORM USING`. Scanning a few applications there seem to be two common ways of
+validating webhook requests:
```

Comment:
Oh, it doesn't need to be anywhere near as tricky as `TRANSFORM USING`! A `MirScalarExpr` is evaluable in any context. We just need to make sure the clause takes a single scalar expression as input, rather than a relation expression.